File: function_for_download/load_nba_data.py
from pathlib import Path
from itertools import product
from urllib.request import urlopen
import tarfile
from typing import Union, Sequence, Optional, List
from io import BytesIO, TextIOWrapper
import csv

import pandas as pd # If use in_memory = True & use_pandas = True


def load_nba_data(path: Union[Path, str] = Path.cwd(),
                  seasons: Union[Sequence, int] = range(1996, 2024),
                  data: Union[Sequence, str] = ("datanba", "nbastats", "pbpstats",
                                                "shotdetail", "cdnnba", "nbastatsv3",
                                                "matchups"),
                  seasontype: str = 'rg',
                  league: str = 'nba',
                  untar: bool = False,
                  in_memory: bool = False,
                  use_pandas: bool = True) -> Optional[Union[List, pd.DataFrame]]:
    """
    Loading a nba play-by-play dataset from github repository https://github.com/shufinskiy/nba_data

    Args:
        path (Union[Path, str]): Path where downloaded file should be saved on the hard disk. Not used if in_memory = True
        seasons (Union[Sequence, int]): Sequence or integer of the year of start of season
        data (Union[Sequence, str]): Sequence or string of data types to load
        seasontype (str): Part of season: rg - Regular Season, po - Playoffs
        league (str): Name league: NBA or WNBA
        untar (bool): Logical: do need to untar loaded archive. Not used if in_memory = True
        in_memory (bool): Logical: If True dataset is loaded into workflow, without saving file to disk
        use_pandas (bool): Logical: If True dataset is loaded how pd.DataFrame, else List[List[str]]. Ignore if in_memory=False

    Returns:
        Optional[pd.DataFrame, List]: If in_memory=True and use_pandas=True return dataset how pd.DataFrame.
        If use_pandas=False return dataset how List[List[str]]
        If in_memory=False return None
    """
    if isinstance(path, str):
        path = Path(path).expanduser()
    if isinstance(seasons, int):
        seasons = (seasons,)
    if isinstance(data, str):
        data = (data,)

    if (len(data) > 1) & in_memory:
        raise ValueError("Parameter in_memory=True available only when loading a single data type")

    if seasontype == 'rg':
        need_data = tuple(["_".join([data, str(season)]) for (data, season) in product(data, seasons)])
    elif seasontype == 'po':
        need_data = tuple(["_".join([data, seasontype, str(season)]) \
                           for (data, seasontype, season) in product(data, (seasontype,), seasons)])
    else:
        need_data_rg = tuple(["_".join([data, str(season)]) for (data, season) in product(data, seasons)])
        need_data_po = tuple(["_".join([data, seasontype, str(season)]) \
                              for (data, seasontype, season) in product(data, ('po',), seasons)])
        need_data = need_data_rg + need_data_po
    if league.lower() == 'wnba':
        need_data = ['wnba_' + x for x in need_data]

    check_data = [file + ".csv" if untar else "tar.xz" for file in need_data]
    not_exists = [not path.joinpath(check_file).is_file() for check_file in check_data]

    need_data = [file for (file, not_exist) in zip(need_data, not_exists) if not_exist]

    with urlopen("https://raw.githubusercontent.com/shufinskiy/nba_data/main/list_data.txt") as f:
        v = f.read().decode('utf-8').strip()

    name_v = [string.split("=")[0] for string in v.split("\n")]
    element_v = [string.split("=")[1] for string in v.split("\n")]

    need_name = [name for name in name_v if name in need_data]
    need_element = [element for (name, element) in zip(name_v, element_v) if name in need_data]

    if in_memory:
        if use_pandas:
            table = pd.DataFrame()
        else:
            table = []
    for i in range(len(need_name)):
        with urlopen(need_element[i]) as response:
            if response.status != 200:
                raise Exception(f"Failed to download file: {response.status}")
            file_content = response.read()
            if in_memory:
                with tarfile.open(fileobj=BytesIO(file_content), mode='r:xz') as tar:
                    csv_file_name = "".join([need_name[i], ".csv"])
                    csv_file = tar.extractfile(csv_file_name)
                    if use_pandas:
                        table = pd.concat([table, pd.read_csv(csv_file)], axis=0, ignore_index=True)
                    else:
                        csv_reader = csv.reader(TextIOWrapper(csv_file, encoding="utf-8"))
                        for row in csv_reader:
                            table.append(row)
            else:
                with path.joinpath("".join([need_name[i], ".tar.xz"])).open(mode='wb') as f:
                    f.write(file_content)
                if untar:
                    with tarfile.open(path.joinpath("".join([need_name[i], ".tar.xz"]))) as f:
                        f.extract("".join([need_name[i], ".csv"]), path)

                    path.joinpath("".join([need_name[i], ".tar.xz"])).unlink()
    if in_memory:
        return table
    else:
        return None

File: function_for_download/load_nba_data.R
#' Loading a dataset from github
#'
#' @description
#' Loading a nba play-by-play dataset from github repository https://github.com/shufinskiy/nba_data
#'
#' @param seasons Sequence or integer of the year of start of season. Not used if in_memory = True
#' @param data Sequence or string of data types to load
#' @param seasontype Part of season: rg - Regular Season, po - Playoffs
#' @param league Name league: NBA or WNBA
#' @param in_memory Upload data without saving to hard disk
#' @param untar do need to untar loaded archive. Not used if in_memory = True
#' @return None
#'
#' @importFrom utils download.file
#' @export
load_nba_data <- function(path = getwd(),
                          seasons = seq(1996, 2024),
                          data = c("datanba", "nbastats", "pbpstats", "shotdetail", 
                                   "cdnnba", "nbastatsv3", "matchups"),
                          seasontype = 'rg',
                          league = 'nba',
                          in_memory = FALSE,
                          untar = FALSE){
  
  path <- normalizePath(path, mustWork = FALSE)
  
  if (length(data) > 1 & in_memory){
    stop("Parameter in_memory=True available only when loading a single data type")
  }
  
  if(seasontype == 'rg'){
    df <- expand.grid(data, seasons)
    need_data <- paste(df$Var1, df$Var2, sep = "_")
  } else if(seasontype == 'po'){
    df <- expand.grid(data, 'po', seasons)
    need_data <- paste(df$Var1, df$Var2, df$Var3, sep = "_")
  } else {
    df_rg <- expand.grid(data, seasons)
    df_po <- expand.grid(data, 'po', seasons)
    need_data <- c(paste(df_rg$Var1, df_rg$Var2, sep = "_"), paste(df_po$Var1, df_po$Var2, df_po$Var3, sep = "_"))
  }
  if(tolower(league) == 'wnba'){
    need_data <- sapply(need_data, function(x){paste0('wnba_', x)}, USE.NAMES = FALSE)
  }
  temp <- tempfile()
  download.file("https://raw.githubusercontent.com/shufinskiy/nba_data/main/list_data.txt", temp)
  f <- readLines(temp)
  unlink(temp)
  v <- unlist(strsplit(f, "="))
  name_v <- v[seq(1, length(v), 2)]
  element_v <- v[seq(2, length(v), 2)]
  need_name <- name_v[which(name_v %in% need_data)]
  need_element <- element_v[which(name_v %in% need_data)]
  
  if(!dir.exists(path)){
    dir.create(path)
  }
  if (in_memory){
    df <- data.frame()
  }
  for(i in seq_along(need_element)){
    if (in_memory){
      temp_file <- tempfile(fileext = ".tar.xz")
      download.file(need_element[i], destfile = temp_file, mode = "wb")
      temp_dir <- tempdir()
      temp_csv <- tempfile(fileext = ".csv", tmpdir = temp_dir)
      untar(temp_file, exdir = temp_dir, files = paste0(
        gsub(".tar.xz", "", basename(need_element[i])),".csv"
      ))
      csv_file <- list.files(temp_dir, pattern = "\\.csv$", full.names = TRUE)
      if (length(csv_file) > 0) {
        tmp_df <- read.csv(csv_file)
        df <- rbind(df, tmp_df)
      } else {
        stop("No CSV file found after extraction.")
      }

      unlink(temp_file)
      unlink(csv_file)
      unlink(temp_dir)
    } else {
      destfile <- paste0(path, '/', need_name[i], ".tar.xz")
      download.file(need_element[i], destfile = destfile)
      if(untar){
        untar(destfile, paste0(need_name[i], ".csv"), exdir = path)
        unlink(destfile)
      }
    }
  }
  if (in_memory){
    return(df)
  }
}
